modules/js-and-css/speculation-rules/helper.php

```diff
@@ -0,0 +1,92 @@
+<?php
+/**
+ * Helper functions used for Speculation Rules.
+ *
+ * @package performance-lab
+ * @since n.e.x.t
+ */
+
+/**
+ * Returns the speculation rules.
+ *
+ * Plugins with features that rely on frontend URLs to exclude from prefetching or prerendering should use the
+ * {@see 'plsr_speculation_rules_href_exclude_paths'} filter to ensure those URL patterns are excluded.
+ *
+ * @since n.e.x.t
+ *
+ * @return array Associative array of speculation rules by type.
+ */
+function plsr_get_speculation_rules() {
+	$prefixer = new PLSR_URL_Pattern_Prefixer();
+
+	$base_href_exclude_paths = array(
+		$prefixer->prefix_path_pattern( '/wp-login.php', 'site' ),
+		$prefixer->prefix_path_pattern( '/wp-admin/*', 'site' ),
```

Comment:
I wonder, is the `PLSR_URL_Pattern_Prefixer` class really needed? Wouldn't this be simpler just doing:

```suggestion
	$base_href_exclude_paths = array(
		site_url( 'wp-login.php', 'login' ),
		admin_url() . '*',
```

Reply:
That's not the same, as these functions return full URLs. The current code only gets the paths.

Reply:
OK, I see it needs to be paths. So then why not:

```suggestion
	$base_href_exclude_paths = array(
		wp_parse_url( site_url( 'wp-login.php', 'login' ), PHP_URL_PATH ),
		wp_parse_url( admin_url(), PHP_URL_PATH ),
```

Reply:
Full URLs are also fine; there's no need to restrict yourself to relative URLs.

Reply:
@westonruter We need to prefix all results from the filter in the same way anyway, so it's IMO a bit pointless to use the WP functions directly here.

Of course we could use the WP functions in all cases, but that would be slower (since for every URL it would have to "recalculate" those base URLs, parse the paths etc.) and more difficult to cover by unit tests.

Reply:
Yes, I suppose it would be a tiny bit slower, but these functions are being used all over the place in WordPress already without any kind of caching so adding two more function calls isn't really going to make a big impact. IMHO, `PLSR_URL_Pattern_Prefixer` seems a bit overkill. Also, what about when a site filters the login URL to be something different using the `login_url` filter? For example, a site that I help maintain has some plugin which filters the login URL to be `.../webadmin/` (for some reason) with `.../wp-login.php` returning a 404. The uses of the prefixer blow could be adapted to use `home_url()`, right?

Reply:
I think this is a bit of a personal preference. Not having this class is more pragmatic, but less testable, and the latter makes it more error-prone (because of less test coverage). The class is simple and decoupled in a way that it can be used easily alongside WordPress's function-driven approach.

Good point regarding the `login_url` filter, it's probably safer to use the `wp_login_url()` function rather than hard-coding `wp-login.php`. Similarly, we should probably use the `admin_url()` function instead of hard-coding `wp-admin`. I'll make that change.